src/interchange/src/confluent.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use anyhow::{bail, Result};
+use byteorder::{BigEndian, ByteOrder};
+
+/// Extracts the schema_id placed in front of the serialized message by the confluent stack
+/// Optionally expect an empty
+///
+/// This function returns the schema_id and a sublice of the rest of the buffer
+pub fn extract_schema_id(buf: &[u8], expect_msg_id: bool) -> Result<(i32, &[u8])> {
+    // The first byte is a magic byte (0) that indicates the Confluent
+    // serialization format version, and the next four bytes are a big
+    // endian 32-bit schema ID.
+    //
+    // https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format
+    //
+    // For formats like protobuf, confluent adds additional information related to
+    // which message in the proto file
+
+    let expected_len = 5 + expect_msg_id as usize;
+
+    if buf.len() < expected_len {
+        bail!(
+            "Confluent-style schema datum is too few bytes: expected at least {} bytes, got {}",
+            expected_len,
+            buf.len()
+        );
+    }
+    let magic = buf[0];
+    let schema_id = BigEndian::read_i32(&buf[1..5]);
+
+    if magic != 0 {
+        bail!(
+            "wrong Confluent-style schema serialization magic: expected 0, got {}",
+            magic
+        );
+    }
+
+    if expect_msg_id {
+        let message_id = buf[5];
+        if message_id != 0 {
+            bail!(
+                "unsupported Confluent-style protobuf message descriptor id: \
+                expected 0, but found: {}. \
+                See https://github.com/MaterializeInc/materialize/issues/9250",
+                message_id
+            );
+        }
+    }
+    Ok((schema_id, &buf[expected_len..]))
+}
```

Comment:
I'd suggest splitting into three functions to make the responsibilities clearer:

```
fn extract_schema_id(buf: &[u8]) -> Result<(i32, &[u8])> {
    // as is
}

pub extract_avro_header(buf: &[u8]) -> Result<(i32, &[u8])> {
    extract_schema_id(buf)
}

pub extract_protobuf_header(buf: &[u8]) -> Result<(i32, &[u8])> {
    let (schema_id, buf) = extract_schema_id(buf)?;
    if buf.is_empty() == 0 { bail!("bad protobuf blah blah") }
    if buf[0] != 0 { bail!("bad protobuf blah blah") }
    Ok((schema_id, buf))
}

```

WDYT? The only cost is one extra buffer length check, but I think it makes it much clearer what error messages can mention "Avro" and what error messages can mention "Protobuf."

Reply:
yah this seems fine, will clean it up a bit